bika/health/upgrade/v01_02_002.py

```diff
@@ -75,3 +87,58 @@ def apply_doctor_permissions_for_clients(portal, ut):
     logger.info(
         "Changed permissions for doctor objects: " +
         "{0}/{1}".format(counter, total))
+
+
+def restore_identifier_types(portal):
+    """Restores the identifier types by using the values stored in patients
+    """
+    logger.info("Restoring additional identifiers for Patients ...")
+
+    # Re-create the Identifier types folder in setup
+    setup = api.get_setup()
+    if "bika_identifiertypes" not in setup:
+        obj = _createObjectByType("IdentifierTypes", setup, "bika_identifiertypes")
+        obj.edit(title="Identifier Types")
+        obj.unmarkCreationFlag()
+        obj.reindexObject()
+
+    # Infer the Identifier Types to be restored from Patients
+    query = dict(portal_type="Patient")
+    brains = api.search(query, CATALOG_PATIENTS)
+    for brain in brains:
+        ids = brain.getPatientIdentifiersStr
+        if ids and ids not in [", "]:
+            # Restore the IdentifierType by using the value stored in Patient
+            obj = api.get_object(brain)
+            restore_identifier_type(obj)
+
+
+def restore_identifier_type(patient):
+    """Restores the identifier types for the Patient passed in
+    """
+    patient_ids = patient.__dict__.get("PatientIdentifiers", [])
+    for patient_id in patient_ids:
+        identifier_id = patient_id.get("IdentifierType")
+        id_value = patient_id.get("Identifier") or patient_id.get("value")
+        if identifier_id and id_value:
+            resolve_identifier_type(identifier_id)
```

Comment:
It would better to get all existing identifier titles outside of the for-loop to check if the ID already exists or not and add eventually created IDs back into that list.